### ISSUE TITLE
How can we optimize the YOLOv8x-OBB model in terms of speed?

### ISSUE THREAD

**Issue by Abhijeet241093** [ultralytics/ultralytics#22055]
### Search before asking

- [x] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/orgs/ultralytics/discussions) and found no similar questions.

### Question

Thank you very much @ultralytics for your support and guidance. We trained different YOLOv8-OBB models (s, m, l, and x series) and exported them into ONNX format. We developed the entire pipeline in C#. However, when we execute the ONNX model, we are unable to achieve the desired speed.

In this case, how can we optimize the speed? Could you suggest different techniques for the same?

**Comment by UltralyticsAssistant** [comment]
👋 Hello @Abhijeet241093, thank you for your interest in Ultralytics 🚀! This is an automated response to help you get support quickly—an Ultralytics engineer will also assist you here soon.

We recommend a visit to the [Docs](https://docs.ultralytics.com/) where you can find many [Python](https://docs.ultralytics.com/usage/python/) and [CLI](https://docs.ultralytics.com/usage/cli/) usage examples and where many of the most common questions may already be answered.

Since this is a custom optimization ❓ Question, please share as much detail as possible to help us understand your setup and provide targeted guidance:
- Your hardware and OS (CPU/GPU model, RAM, OS version)
- ONNX Runtime or inference backend (providers enabled, versions)
- Export details (export command, opset, dynamic/static shapes, input image size)
- Model variants used (s/m/l/x) and any post-processing steps (NMS settings, OBB decoding)
- Batch size, preprocessing pipeline, and I/O formats
- Timing breakdown if available (preprocess/inference/postprocess) and minimal logs
- A few representative input images and expected outputs

If this is a 🐛 Bug Report, please provide a [minimum reproducible example](https://docs.ultralytics.com/help/minimum-reproducible-example/) to help us debug it.

Join the Ultralytics community where it suits you best. For real-time chat, head to [Discord](https://discord.com/invite/ultralytics) 🎧. Prefer in-depth discussions? Check out [Discourse](https://community.ultralytics.com/). Or dive into threads on our [Subreddit](https://www.reddit.com/r/Ultralytics/) to share knowledge with the community.

## Upgrade

Upgrade to the latest `ultralytics` package including all [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml) in a [Python>=3.8](https://www.python.org/) environment with [PyTorch>=1.8](https://pytorch.org/get-started/locally/) to verify your issue is not already resolved in the latest version:

```bash
pip install -U ultralytics
```

## Environments

YOLO may be run in any of the following up-to-date verified environments (with all dependencies including [CUDA](https://developer.nvidia.com/cuda-zone)/[CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/) and [PyTorch](https://pytorch.org/) preinstalled):

- Notebooks with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo11"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- Google Cloud Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
- Amazon Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
- Docker Image. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>

If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are currently passing. CI tests verify correct operation of all YOLO [Modes](https://docs.ultralytics.com/modes/) and [Tasks](https://docs.ultralytics.com/tasks/) on macOS, Windows, and Ubuntu every 24 hours and on every commit.

Once you share the details above, we’ll help narrow down the bottlenecks and next steps ⚡️

**Comment by Y-T-G** [comment]
If you have NVIDIA GPU, you should use TensorRT which would be much faster than ONNX. Otherwise you can try using OpenVINO if you're running on CPU.

https://docs.ultralytics.com/modes/export/

You can export for with `half=True`. Or use INT8 for even stronger quantization, but with a reduction in accuracy.

Regardless, YOLOv8x is one of the larges YOLO models, and it's not going to be very fast. You should use a smaller model if you are interested in faster speed.

**Comment by Abhijeet241093** [comment]
Hello Sir, the problem is that TensorRT is not supported in C#, so we are searching for other options. We have tried the s, m, and l series models, but there is still a speed issue. Can you suggest other possible option ? that can be implemented in onnx model ?

**Comment by Y-T-G** [comment]
You can write the inference code in C++ for TensorRT, create a DLL and then use that in your C# application.

**Comment by Abhijeet241093** [comment]
That's good idea. Can you please suggest reference link regarding your suggtion ?

**Comment by glenn-jocher** [comment]
You don’t need a custom C++ DLL—ONNX Runtime’s C# API can use the TensorRT Execution Provider; build ORT with TensorRT and the C# bindings, then enable it like this: 
```csharp
using Microsoft.ML.OnnxRuntime;

var so = SessionOptions.MakeSessionOptionWithTensorrtProvider(deviceId: 0);
so.GraphOptimizationLevel = GraphOptimizationLevel.ORT_ENABLE_ALL;  // optional
so.AppendExecutionProvider_CUDA(0);  // optional fallback for unsupported ops

using var session = new InferenceSession("yolov8x-obb.onnx", so);
```
Docs: see the ONNX Runtime TensorRT EP guide, the C# API reference for `SessionOptions.AppendExecutionProvider_Tensorrt`, and how to build ORT with EPs (create a custom NuGet with `--build_csharp`). If you’re on Intel CPU/GPU, the OpenVINO EP for C# or exporting to OpenVINO are solid options; tuning tips are here. ([onnxruntime.ai](https://onnxruntime.ai/docs/execution-providers/TensorRT-ExecutionProvider.html))
